### PR TITLE
Improve error handling in cucumber (esp. 'visit_page')

### PR DESCRIPTION
**Pull request by weedySeaDragon**
### PT Story:  improve cuke error handling
https://www.pivotaltracker.com/story/show/166369559

The current error handling in  the
`Given(/^I am on the "([^"]*)" page(?: for "([^"]*)")?$/) do |page, email|`  step in `features\step_definitions\navigation.rb` is "swallowing" errors.  That is, if an error/exception is raised that is not one explicitly rescued, information about it is lost.   

In order to rework this, I needed to (1) create explicit Error classes that can be rescued, and (2) test the code (obv).  In order to test this, I pulled out the code into separate `modules`.  Then wrote some RSpecs.  While this verges on overkill, I really did have to do most of it in order to verify that it was working (== testing).

Plus the RSpecs show how the classes can be used.  _Puttin' the "Spec" in RSpec_

### Changes proposed in this pull request:
1.  Pulled code out into explicit modules and separate files.  Moved them into `/support`
2. Created error classes:
   - `SHFStepsError` as the main parent class.  If there is already an exception message that has been raised (in `$!`), this will put it into an attribute so that it can be displayed.  This helps ensure that an original exception won't be "swallowed" and disappear.

3. RSpecs for the error classes and for the error handling in `visit_page`:
   - `spec/features/support/cucumber_steps_errors_spec.rb`
   - `spec/features/support/visit_page_spec.rb`

### Ready for review:
@patmbolger

### CODE REVIEW

File: spec/features/support/cucumber_steps_errors_spec.rb [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+      begin
+        raise error_class
+      rescue NoMethodError, error_class
+        shf_error = error_class.new
+      end
+
+      expect(shf_error.message).to eq "SHFStepsError"
+    end
+
+    it 'shows just the class nameif it was the most recent error rasied' do
```
Comment (patmbolger) [inline spec/features/support/cucumber_steps_errors_spec.rb]: couple of spelling errors ...

File: features/support/step_errors.rb [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+  attr_reader :original_error_message
+
+  # If there is an error message in the global variable '$!' then this will
+  # use it. This enables the object to “magically” discover the originating
+  # exception message on initialization.
+  def initialize
+    super
+    @original_error_message = $! # This is the message of the last raised exception
+  end
```
Comment (patmbolger) [inline features/support/step_errors.rb]: I suggest considering use of `Exception#cause` as a way to get information related to predecessor exceptions.  

This is cleaner (IMO) and leverages Ruby better.  It also allows for gathering information for any number of levels of predecessor exceptions.

I opened a PR - against your PR - where I made this change (and adjusted tests accordingly).

Your way here works fine, of course, but we are also establishing a pattern on how to do more sophisticated exception management, and as such this PR is important (I did happen to use `cause` for payments errors).

Here's a great (and short) write up on this: https://www.honeybadger.io/blog/nested-errors-in-ruby-with-exception-cause/

**Review by patmbolger — COMMENTED** [review]
Please see comment.

File: features/support/step_errors.rb (hunk 2 above)
Comment (weedySeaDragon) [reply features/support/step_errors.rb]: Excellent suggestion.  I totally agree: we want this to work really well and `cause` is a better approach.  (Starr has posted lots of great info.  And they started Honeybadger in the neighborhood where I used to live. :-) )

**Review by patmbolger — APPROVED** [review]
Looks good!

**Pull request merged**